modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java

```diff
@@ -820,6 +830,27 @@ private void enlist(SourceAwareIgniteRel rel) {
             }.visit(fragment.root());
         }
 
+        private CompletableFuture<MultiStepPlan> mapFragments(MultiStepPlan plan) {
+            List<IgniteRel> fragments = plan.fragments()
+                    .stream()
+                    .map(Fragment::root)
+                    .collect(Collectors.toList());
```

Comment:
Does it make sense to avoid collection copying?

We could use smth like `org.apache.ignite.internal.utilTransformingIterator`

Reply:
Sure. I thought of it but I forgot about `TransformingIterator`.

Reply:
Well, here is a problem it is a bad idea to accept an iterator because is modifiable. We need something like transforming iterable here.

Reply:
Added `TransformingIterator::newIterable`. And replaced `List<IgniteRel>` with `Iterable<IgniteRel>` in DepedencyResolver's signature.